pkg/adsc/adsc.go

```diff
@@ -539,25 +614,55 @@ func (a *ADSC) handleRecv() {
 
 		// TODO: add hook to inject nacks
 
-		a.mutex.Lock()
-		if len(gvk) == 3 {
-			gt := resource.GroupVersionKind{Group: gvk[0], Version: gvk[1], Kind: gvk[2]}
-			a.sync[gt.String()] = time.Now()
-		}
-		a.Received[msg.TypeUrl] = msg
-		a.ack(msg)
-		a.mutex.Unlock()
+		a.onReceive(msg)
 
 		if len(listeners) > 0 {
+			if a.LocalCacheDir != "" {
+				strResponse, err := json.MarshalIndent(listeners, "  ", "  ")
+				if err == nil {
+					err = ioutil.WriteFile(a.LocalCacheDir+"_lds.json", strResponse, 0644)
+					if err != nil {
+						log.Warna("Failed to save ", err)
+					}
+				}
+			}
+
 			a.handleLDS(listeners)
 		}
 		if len(clusters) > 0 {
+			if a.LocalCacheDir != "" {
+				strResponse, err := json.MarshalIndent(clusters, "  ", "  ")
+				if err == nil {
+					err = ioutil.WriteFile(a.LocalCacheDir+"_cds.json", strResponse, 0644)
+					if err != nil {
+						log.Warna("Failed to save ", err)
+					}
+				}
+			}
 			a.handleCDS(clusters)
 		}
 		if len(eds) > 0 {
+			if a.LocalCacheDir != "" {
+				strResponse, err := json.MarshalIndent(eds, "  ", "  ")
```

Comment:
Cannot use `json` library, should be `jsonpb`

Reply:
Also can you break this out so we don't repeat these 10 lines for each resource?

Reply:
Moved to function. I think we had a discussion about jsonbp - not sure it works for both gogo and google. 
For debugging json seems to work pretty well - do you remember why we have to use jsonpb ? 

Again - may become a problem when we want to reload - but I'm thinking to just save binary protos for that and avoid all the crazyness.